### PR TITLE
Fix #1933:  Use FrozenArray for AudioWorkletProcessor process()

### LINKED ISSUE #1933: Improve description of parameters for process() calls on client objects extending AudioWorkletProcessor
https://webaudio.github.io/web-audio-api/#rendering-loop says 'Let `processFunction` be the result of a `Get(O=processor, P="process")`', and so `processFunction` is an ECMAScript entity.

The `process()` method is not defined as WebIDL callback type, but its parameters appear to have WebIDL types.
https://webaudio.github.io/web-audio-api/#dom-audioworkletprocessor-process-inputs-outputs-parameters-inputs says that `inputs` has type `sequence<sequence<Float32Array>>` and links to the WebIDL type https://heycam.github.io/webidl/#idl-Float32Array

https://webaudio.github.io/web-audio-api/#dom-audioworkletprocessor-process-inputs-outputs-parameters-parameters says `parameters` is "A map of string keys and associated Float32Arrays", also linking to https://heycam.github.io/webidl/#idl-Float32Array.
Is this https://heycam.github.io/webidl/#idl-record ?

### PR DESCRIPTION
**Pull request by rtoy**
Instead of `sequence<sequence<Float32Array>>`, use `FrozenArray<FrozenArray<Float32Array>>`.

Define a callback function as well.

***
<a href="https://pr-preview.s3.amazonaws.com/rtoy/web-audio-api/pull/2250.html" title="Last updated on Nov 6, 2020, 10:31 PM UTC (ae984bd)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/WebAudio/web-audio-api/2250/d50c68f...rtoy:ae984bd.html" title="Last updated on Nov 6, 2020, 10:31 PM UTC (ae984bd)">Diff</a>

### CODE REVIEW

**Comment by rtoy** [comment]
Note to self:  Making `process()` be a callback is easy.  Just not sure how to integrate that into the spec appropriately.

**Comment by rtoy** [comment]
> bikeshed: `AudioWorkletProcessCallback` would be a more specific name, which might be useful should any other `AudioWorkletProcessor` callbacks be added in the future.

That works for me.

**Comment by rtoy** [comment]
> The third parameter has named properties rather than indexed properties and so is not an array. It is a frozen object, but there is no such pre-defined WebIDL type. It can be simply `object` here, with details described in the prose.

Yes, my mistake. Too much search-and-replace.  I'll have to ask @hoch exactly how he did this.

**Comment by rtoy** [comment]
I think the latest revision is closer to what we want, but I'm not 100% sure.  Still have to describe what happens with all of the FrozenArrays and the callback parameters object map if you do something bad.  But I hope this is close enough that we can now concentrate on getting the right text for that.

File: index.bs [hunk 1]
```diff
@@ -10454,86 +10462,62 @@ Constructors</h5>
...
-			any {{AudioWorkletProcessor}} that is active only when it has
-			active inputs.
-		</div>
-
-		<pre class=argumentdef for="AudioWorkletProcessor/process(inputs, outputs, parameters))">
-			inputs:
- 				The input audio buffer from the incoming connections provided by the user agent. It has type <code>sequence&lt;sequence&lt;Float32Array>></code>.<code class="nobreak">inputs[n][m]</code> is a {{Float32Array}} of audio samples for the \(m\)th channel of the \(n\)th input. While the number of inputs is fixed at construction, the number of channels can be changed dynamically based on [=computedNumberOfChannels=].
-
-				If there are no [=actively processing=] {{AudioNode}}s connected to the \(n\)th input of the {{AudioWorkletNode}} for the current render quantum, then the content of <code>inputs[n]</code> is an empty array, indicating that zero channels of input are available. This is the only circumstance under which the number of elements of <code>inputs[n]</code> can be zero.
+The [=process()=] callback function is called
```
Comment (hoch) [inline index.bs:10475]: Instead of explaining how this is invoked here, referring https://webaudio.github.io/web-audio-api/#rendering-loop directly might be clearer and more specific.

File: index.bs [hunk 2]
```diff
@@ -10568,6 +10552,28 @@ class MyProcessor extends AudioWorkletProcessor {
 }
 </xmp>
 
+<h6 id="audioworkletprocess-callback-parameters" callback lt="AudioWorkletProcessCallback()">
+Callback {{AudioWorkletProcessCallback}} Parameters
+</h6>
+The {{AudioWorkletProcessCallback}} function has the following parameters:
+
+*Insert text here about how these FrozenArrays behave if they get transferred*.
```
Comment (hoch) [inline index.bs:10560]: It will get reallocated when the array topology changes or it is transferred.

**Review by hoch — COMMENTED** [review]
I am not sure how to make a connection between `AudioWorkletProcessCallback` to `AudioWorkletProcessor.process()`.

File: index.bs [hunk 3]
```diff
@@ -43,7 +43,10 @@ Markup Shorthands: markdown on, dfn on, css off
 spec: ECMAScript; url: https://tc39.github.io/ecma262/#sec-data-blocks; type: dfn; text: data block;
 url: https://www.w3.org/TR/mediacapture-streams/#dom-mediadevices-getusermedia; type: method; for: MediaDevices; text: getUserMedia()
 </pre>
-
+<!-- We want IDL object -->
```
[the remark this answers is not in the document]
Comment (rtoy) [reply index.bs:46]: Oh,  bikeshed was choosing something else when we did `{{object}}`.  This makes it link to `object` in the webidl spec, which is want we want, I think.

File: index.bs [hunk 4]
```diff
@@ -10356,13 +10359,18 @@ and the definition of the class manifests the actual audio processing.
...
 interface AudioWorkletProcessor {
 	constructor ();
 	readonly attribute MessagePort port;
 };
-</pre>
+
+callback AudioWorkletProcessCallback =
+  boolean (FrozenArray<FrozenArray<Float32Array>> inputs,
+           FrozenArray<FrozenArray<Float32Array>> outputs,
+           object  parameters);
```
[the remark this answers is not in the document]
Comment (rtoy) [reply index.bs:10372]: Fixed.

File: index.bs [hunk 5]
```diff
@@ -10373,7 +10381,7 @@ interface AudioWorkletProcessor {
 
 	: <dfn>[[callable process]]</dfn>
 	::
-		A boolean flag representing whether {{AudioWorkletProcessor/process()}} is
+		A boolean flag representing whether [=process()=]
```
[the remark this answers is not in the document]
Comment (rtoy) [reply index.bs:10384]: Fixed.

File: index.bs (hunk 1 above)
Comment (rtoy) [reply index.bs:10475]: This was pretty much in the original, but I can reference the rendering loop.

File: index.bs (hunk 2 above)
Comment (rtoy) [reply index.bs]: Added some new text here.  PTAL.

File: index.bs (hunk 1 above)
Comment (hoch) [reply index.bs:10475]: Yeah, I think referencing the algorithm itself would be more thorough.

File: index.bs [hunk 6]
```diff
@@ -10454,86 +10462,62 @@ Constructors</h5>
...
-			outputs:
-				The output audio buffer that is to be consumed by the user agent. It has type <code>sequence&lt;sequence&lt;Float32Array>></code>.<code class="nobreak">outputs[n][m]</code> is a {{Float32Array}} object containing the audio samples for \(m\)th channel of \(n\)th output. Each of the {{Float32Array}}s are zero-filled. The number of channels in the output will match [=computedNumberOfChannels=] only when the node has a single output.
+The [=process()=] callback function is called
+synchronously by the audio <a>rendering thread</a> at
+every <a>render quantum</a>, if {{AudioWorkletNode}} is
+[=actively processing=].
 
-			parameters:
-				An [=ordered map=] of <var>name</var> → <var>parameterValues</var>. <code>parameters["<var>name</var>"]</code> returns <var>parameterValues</var>, which is a {{Float32Array}} with the automation values of the <var>name</var> {{AudioParam}}.
+This callback returns a {{boolean}}.
```
Comment (hoch) [inline index.bs:10480]: The IDL shows the return type. Not sure if this is needed.

File: index.bs [hunk 7]
```diff
@@ -10568,6 +10552,39 @@ class MyProcessor extends AudioWorkletProcessor {
...
+	:: The output audio buffer that is to be consumed by the user agent. <code class="nobreak">outputs[n][m]</code> is a {{Float32Array}} object containing the audio samples for \(m\)th channel of \(n\)th output. Each of the {{Float32Array}}s are zero-filled. The number of channels in the output will match [=computedNumberOfChannels=] only when the node has a single output.
+
+	: {{AudioWorkletProcessCallback/parameters!!argument}}, of type {{object}}
+	:: An [=ordered map=] of <var>name</var> → <var>parameterValues</var>. <code>parameters["<var>name</var>"]</code> returns <var>parameterValues</var>, which is a {{FrozenArray}}&lt;{{Float32Array}}> with the automation values of the <var>name</var> {{AudioParam}}.
+
+		For each array, the array contains the [=computedValue=] of the parameter for all frames in the [=render quantum=]. However, if no automation is scheduled during this render quantum, the array MAY have length 1 with the array element being the constant value of the {{AudioParam}} for the [=render quantum=].
+
+		*Insert text here about how these aren't freshly allocated each time, unless the object is transferred.*
```
Comment (hoch) [inline index.bs:10584]: This still says "Insert text here...".

File: index.bs [hunk 8]
```diff
@@ -10568,6 +10552,39 @@ class MyProcessor extends AudioWorkletProcessor {
...
+
+In general, the {{AudioWorkletProcessCallback/inputs!!argument}} and
+{{AudioWorkletProcessCallback/outputs!!argument}} arrays will be reused
+between calls so that no memory allocation is done.  However, if the
+topology changes, because, say, the number of channels in the input or the
+output changes, new arrays are reallocated.  New arrays are also
+reallocated if any part of the
+{{AudioWorkletProcessCallback/inputs!!argument}} or
+{{AudioWorkletProcessCallback/outputs!!argument}} arrays are
+transferred.
```
Comment (hoch) [inline index.bs:10568]: This doesn't include `parameters` object. We can (and Chrome does) freeze the object and reallocate if the topology changes.

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/freeze
https://source.chromium.org/chromium/chromium/src/+/master:third_party/blink/renderer/modules/webaudio/audio_worklet_processor.cc;l=460

File: index.bs (hunk 1 above)
Comment (rtoy) [reply index.bs:10475]: I will remove this and just reference the algorithm.

File: index.bs (hunk 7 above)
Comment (rtoy) [reply index.bs:10584]: A reminder that this still needs to be fixed.  Either here or in the part about inputs/outputs.

File: index.bs (hunk 8 above)
Comment (rtoy) [reply index.bs:10568]: IIRC, the comments in #1933 says we'd have to do much more work than just saying "Freeze" because there's no WebIDL concept of freeze other than FrozenArrays.

File: index.bs (hunk 6 above)
Comment (rtoy) [reply index.bs:10480]: Yes, this should be removed.

File: index.bs (hunk 8 above)
Comment (hoch) [reply index.bs:10568]: Correct. It's less ideal, but we can directly refer ECMAScript algorithm:
https://tc39.es/ecma262/#sec-object.freeze

Or start from WebIDL's FrozenArray's algorithm:
https://heycam.github.io/webidl/#es-frozen-array

**Comment by rtoy** [comment]
Question:  Should we add a something like Example 38 in https://heycam.github.io/webidl/#idl-callback-functions?  That is add a method, say, `performOp(AudioWorkletProcessCallback process)`?

This is a pretty big change in interface so perhaps that's too late for that.

**Comment by karlt** [comment]
> Question: Should we add a something like Example 38 in https://heycam.github.io/webidl/#idl-callback-functions? That is add a method, say, `performOp(AudioWorkletProcessCallback process)`?

That shouldn't be necessary. `performOperation` in Example 38 identifies the callback, but for `AudioWorkletProcessCallback`, this is provided by the [`process` property on the registered processor](https://webaudio.github.io/web-audio-api/#rendering-loop).

**Comment by rtoy** [comment]
> > Question: Should we add a something like Example 38 in https://heycam.github.io/webidl/#idl-callback-functions? That is add a method, say, `performOp(AudioWorkletProcessCallback process)`?
> 
> That shouldn't be necessary. `performOperation` in Example 38 identifies the callback, but for `AudioWorkletProcessCallback`, this is provided by the [`process` property on the registered processor](https://webaudio.github.io/web-audio-api/#rendering-loop).

Works for me.  Thanks!

**Comment by rtoy** [comment]
PTAL. I think I have all the pieces in place now.  The preview diff is kind of hard to follow.  Perhaps using that and the preview would make it easier to see what was changed.

**Review by hoch — APPROVED** [review]
LGTM
